pkg/deployments/scaler.go

```diff
@@ -89,19 +89,22 @@ func (s *scaler) compareScales(current, desired int32) {
 	if s.desiredScale > s.currentScale {
 		// Scale up immediately.
 		go s.scaleFunc(s.desiredScale, false)
-		s.scaleDownStarted = false
+		s.stopScaleDown()
 	} else if s.desiredScale == s.currentScale {
 		// Do nothing, schedule nothing.
-		if s.scaleDownTimer != nil {
-			s.scaleDownTimer.Stop()
-		}
-		s.scaleDownStarted = false
+		s.stopScaleDown()
 	} else {
 		// Schedule a scale down.
-
 		if s.scaleDownTimer == nil {
 			s.scaleDownTimer = time.AfterFunc(s.scaleDownDelay, func() {
-				if err := s.scaleFunc(s.desiredScale, false); err != nil {
+				s.mtx.Lock()
+				s.scaleDownStarted = false // mark completed already
+				desiredScale, currentScale := s.desiredScale, s.currentScale
+				s.mtx.Unlock()
+				if desiredScale == -1 || desiredScale == currentScale {
```

Comment:
s.desiredScale could only be `-1` right after initialization right? because there is this if earlier in scaler:

```
	if current != -1 {
		s.currentScale = current
	}
	if desired != -1 {
		s.desiredScale = desired
	}
```

asking to improve my understanding, this is not feedback to your code or a request for any changes